File: benchmarks/benchmark_vs_tcrdist3.py
#!/usr/bin/env python3
"""
Benchmark Python tcrdist3 for comparison with tcrdistR.

Usage: python benchmarks/benchmark_vs_tcrdist3.py [path_to_dash.csv]

Requires: pip install tcrdist3 pandas
"""

import sys
import time
import pandas as pd
from tcrdist.repertoire import TCRrep


def main():
    if len(sys.argv) >= 2:
        dash_file = sys.argv[1]
    else:
        dash_file = "tests/testthat/fixtures/dash.csv"

    print(f"Loading data from: {dash_file}")
    dash = pd.read_csv(dash_file)
    dash = dash.rename(columns={
        "v_a_gene": "va",
        "cdr3_a_aa": "cdr3a",
        "v_b_gene": "vb",
        "cdr3_b_aa": "cdr3b",
    })
    print(f"Total TCRs: {len(dash)}\n")

    sizes = [50, 100, 200, 500, 1000, len(dash)]
    sizes = [s for s in sizes if s <= len(dash)]

    results = []
    for n in sizes:
        sub = dash.head(n).copy().reset_index(drop=True)
        npairs = n * (n - 1) / 2

        # tcrdist3 computes the matrix via TCRrep
        sub_rep = sub[["va", "cdr3a", "vb", "cdr3b"]].copy()
        sub_rep["count"] = 1

        t0 = time.time()
        tr = TCRrep(
            cell_df=sub_rep,
            organism="mouse",
            chains=["alpha", "beta"],
            compute_distances=True,
        )
        elapsed = time.time() - t0

        rate = npairs / elapsed if elapsed > 0 else float("inf")
        results.append((n, npairs, elapsed, rate))
        print(f"  n={n:5d} | {npairs:10.0f} pairs | {elapsed:7.4f} sec | {rate:10.0f} pairs/sec")

    print("\n--- tcrdist3 (Python) Benchmark Results ---")
    print(f"{'n':>6}  {'pairs':>12}  {'time_sec':>10}  {'pairs/sec':>12}")
    for n, npairs, elapsed, rate in results:
        print(f"{n:6d}  {npairs:12.0f}  {elapsed:10.4f}  {rate:12.0f}")

    import tcrdist
    print(f"\nPython version: {sys.version.split()[0]}")
    print(f"tcrdist3 version: {tcrdist.__version__}")


if __name__ == "__main__":
    main()

File: benchmarks/benchmark_vs_tcrdist3.R
#!/usr/bin/env Rscript
#
# Benchmark tcrdistR vs Python tcrdist3
#
# Usage: Rscript benchmarks/benchmark_vs_tcrdist3.R [path_to_dash.csv]
#
# If no path is given, looks for tests/testthat/fixtures/dash.csv in the
# package source tree (assumes running from the package root).

library(tcrdistR)

args <- commandArgs(trailingOnly = TRUE)
if (length(args) >= 1L) {
    dash_file <- args[1L]
} else {
    dash_file <- "tests/testthat/fixtures/dash.csv"
}

if (!file.exists(dash_file)) {
    stop("Cannot find dash.csv at: ", dash_file,
         "\nRun from the package root or provide the path as an argument.")
}

cat("Loading data from:", dash_file, "\n")
dash <- utils::read.csv(dash_file, stringsAsFactors = FALSE)
dash <- data.frame(
    va    = dash$v_a_gene,
    cdr3a = dash$cdr3_a_aa,
    vb    = dash$v_b_gene,
    cdr3b = dash$cdr3_b_aa,
    stringsAsFactors = FALSE
)
cat("Total TCRs:", nrow(dash), "\n\n")

# Warm up (gene database loading, JIT)
invisible(tcrdist_matrix(dash[1:10, ], "mouse"))

sizes <- c(50L, 100L, 200L, 500L, 1000L, nrow(dash))
sizes <- sizes[sizes <= nrow(dash)]

results <- data.frame(
    n = integer(0),
    pairs = numeric(0),
    time_sec = numeric(0),
    pairs_per_sec = numeric(0)
)

for (n in sizes) {
    sub <- dash[seq_len(n), ]
    npairs <- as.double(n) * (n - 1) / 2

    timing <- system.time({
        mat <- tcrdist_matrix(sub, "mouse")
    })

    elapsed <- timing["elapsed"]
    rate <- npairs / elapsed

    results <- rbind(results, data.frame(
        n = n,
        pairs = npairs,
        time_sec = round(elapsed, 4),
        pairs_per_sec = round(rate, 0)
    ))

    cat(sprintf("  n=%5d | %10.0f pairs | %7.4f sec | %10.0f pairs/sec\n",
                n, npairs, elapsed, rate))
}

cat("\n--- tcrdistR Benchmark Results ---\n")
print(results, row.names = FALSE)
cat("\nR version:", R.version.string, "\n")
cat("tcrdistR version:", as.character(packageVersion("tcrdistR")), "\n")
